nixos/modules/security/selinux.nix

```diff
@@ -0,0 +1,119 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.security.selinux;
+in
+{
+  meta.maintainers = with lib.maintainers; [
+    RossComputerGuy
+  ];
+
+  options.security.selinux = {
+    enable = lib.mkEnableOption "SELinux";
+    policy = lib.mkOption {
+      type = lib.types.path;
+      description = "The path to the SELinux policy";
+      defaultText = lib.literalExpression ''"''${pkgs.selinux-refpolicy.override { inherit (config.security.selinux) policyVersion; }}/share/selinux/refpolicy"'';
+      default = "${
+        pkgs.selinux-refpolicy.override {
+          inherit (config.security.selinux) policyVersion;
+        }
+      }/share/selinux/refpolicy";
+    };
+    policyVersion = lib.mkOption {
+      type = lib.types.nullOr lib.types.int;
+      description = "The version of the SELinux policy";
+      default = null;
+    };
+    type = lib.mkOption {
+      type = lib.types.str;
+      description = "The SELinux policy type to load";
+      default = "refpolicy";
+    };
+    mode = lib.mkOption {
+      type = lib.types.enum [
+        "enforcing"
+        "permissive"
+        "disabled"
+      ];
+      description = "The enforcement mode";
+      default = "permissive";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    boot = {
+      kernelPatches = [
+        {
+          name = "selinux";
+          extraStructuredConfig = with lib.kernel; {
+            SECURITY_SELINUX = yes;
+            SECURITY_SELINUX_BOOTPARAM = yes;
+          };
+          patch = null;
+        }
+      ];
+      kernelParams = [ "security=selinux" ];
+    };
+
+    system.activationScripts.selinux = {
+      deps = [ "etc" ];
+      text = ''
+        install -d -m0755 /var/lib/selinux
+        cmd="${lib.getExe' pkgs.policycoreutils "semodule"} -s ${cfg.type} -i ${cfg.policy}/*.pp"
+        skipSELinuxActivation=0
+
+        if [ -e /var/lib/selinux/activate-check ]; then
+          if [ "$(cat /var/lib/selinux/activate-check)" == "$cmd" ]; then
+            skipSELinuxActivation=1
```

Comment:
Do we want to run this every boot?

Reply:
Only when things change and first boot, that's why it's set up like this.

Reply:
Have you looked into:

```
   --refresh
              Like --build, but reuses existing linked policy if no
              changes to module files are detected (by comparing with
              checksum from the last transaction).  One can use this
              instead of -B to ensure that any changes to the module
              store done by an external tool (e.g. a package manager) are
              applied, while automatically skipping the module re-linking
              if there are no module changes.
```

Reply:
No but does it work when no modules are installed? Or would we need to do `-i` still?

Reply:
-i takes a module package. To the point made above, you can do multiple -i for whatever module packages you want, or manually mess with the filesystem to assemble them, then --refresh syncs changes. So because we are doing -i I don't think we need --refresh.

At least that is my current read. But, also by that current read, we may need to remove the old modules since -i installs or replaces, and the store path may change and we won't clean up all the old .pp files.

Reply:
Maybe the activate check file could be a line separated list of modules we activated? (i.e. the result of ls ${module}/*). Then if it changes at all, we can remove all the old ones and add all the new ones.

Reply:
Is there a way to completely clear out the old policy, then? My point is that if there's a SELinux policy module loaded in the old policy but not in the new one, it will still be loaded after reactivation. I believe this is a valid concern and not a hypothetical, so was offering some ways around it. We should also make sure that the same policy modules are replaced after reactivation and not duplicated with the old ones.

Reply:
I think we'll have to `rm -rf /etc/selinux/$policy`.

Reply:
I also wonder if we could compile the policy (all those bb files) into a derivation instead of at activation... might make things faster since we have a whole build step to take advantage of over Red Hat.

Reply:
Yeah, "baking" it as a derivation is the preferred way imo. However, I think the implementation I have is a decent enough to get it working. This won't be the last iteration of this module. I think we can focus on the store stuff rather than activation right now.